### PR TITLE
Mte inventory block

### PR DESCRIPTION
**Pull request by BlueWeabo**
Some work on MuTEs. We are now capable of expanding the inventory.

### CODE REVIEW

File: src/main/java/gregtech/api/multitileentity/machine/MultiTileBasicMachine.java [hunk 1]
```diff
@@ -36,8 +36,8 @@ public class MultiTileBasicMachine extends BaseTickableMultiTileEntity {
     protected FluidTankGT[] mTanksInput = GT_Values.emptyFluidTankGT, mTanksOutput = GT_Values.emptyFluidTankGT;
     protected FluidStack[] mOutputFluids = GT_Values.emptyFluidStack;
 
-    protected final IItemHandlerModifiable mInputInventory = new ItemStackHandler(17);
-    protected final IItemHandlerModifiable mOutputInventory = new ItemStackHandler(15);
+    protected final IItemHandlerModifiable mInputInventory = new ItemStackHandler(16);
```
Comment (mitchej123) [inline src/main/java/gregtech/api/multitileentity/machine/MultiTileBasicMachine.java:39]: These should probably be put back to empty inventory, and the constructor should create the inventory based on the NBT size for that multi (similar to the materials).

File: src/main/java/gregtech/api/multitileentity/multiblock/base/MultiBlockController.java [hunk 2]
```diff
@@ -551,6 +553,10 @@ public boolean check(S t, World world, int x, int y, int z) {
                 if (tTarget != null && tTarget != MultiBlockController.this) return false;
 
                 part.setTarget(MultiBlockController.this, aModes);
+
+                if (part instanceof AdvancedCasing) {
+                    mUpgradeCasings.add((AdvancedCasing) part);
```
Comment (mitchej123) [inline src/main/java/gregtech/api/multitileentity/multiblock/base/MultiBlockController.java:558]: I'd probably make this into a function `registerCasing()` so we can have it do anything else it needs to at registration time, potentially including a callback to the casing itself... 🤔

File: src/main/java/gregtech/api/multitileentity/multiblock/base/MultiBlockPart.java [hunk 3]
```diff
@@ -208,9 +216,21 @@ public void writeMultiTileNBT(NBTTagCompound aNBT) {
...
+    @Override
+    public int getLockedInventoryIndex() {
+        return mLockedInventoryIndex;
+    }
+
     @Override
     public void setTargetPos(ChunkCoordinates aTargetPos) {
         mTargetPos = aTargetPos;
+        IMultiBlockController mTarget = getTarget(false);
+        setTarget(mTarget, mAllowedModes);
```
Comment (mitchej123) [inline src/main/java/gregtech/api/multitileentity/multiblock/base/MultiBlockPart.java:233]: Why was this needed?

File: src/main/java/gregtech/api/multitileentity/multiblock/casing/InventoryUpgrade.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.nbt.NBTTagList;
+
+public class InventoryUpgrade extends AdvancedCasing {
+    public static final int INPUT = 0;
+    public static final int OUTPUT = 1;
+    private String mInventoryName = "inventory";
+    private int mInventorySize = 16;
+    private IItemHandlerModifiable mInputInventory;
```
Comment (mitchej123) [inline src/main/java/gregtech/api/multitileentity/multiblock/casing/InventoryUpgrade.java:18]: Hmm.  My original intention was the inventories would all be stored in the controller, this casing would just tell the controller to create a new inventory, with a key, linked to this casing.  The controller would then be responsible for all saving/loading/handling of it.

File: src/main/java/gregtech/api/multitileentity/multiblock/base/MultiBlockController.java [hunk 5]
```diff
@@ -1062,6 +1092,11 @@ public void addGregTechLogo(ModularWindow.Builder builder) {
 
     @Override
     public void addUIWidgets(ModularWindow.Builder builder, UIBuildContext buildContext) {
+        if (isServerSide()) {
+            for (AdvancedCasing tPart : mUpgradeCasings) {
+                tPart.issueClientUpdate();
```
Comment (mitchej123) [inline src/main/java/gregtech/api/multitileentity/multiblock/base/MultiBlockController.java:1097]: I think this goes away once the inventory lives on the controller.

File: src/main/java/gregtech/api/multitileentity/multiblock/base/MultiBlockPart.java (hunk 3 above)
Comment (BlueWeabo) [reply src/main/java/gregtech/api/multitileentity/multiblock/base/MultiBlockPart.java:233]: so that the inventory can sync between server and client. As on setTarget is when the block registers its inventory to the multi

File: src/main/java/gregtech/api/multitileentity/multiblock/casing/InventoryUpgrade.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+    @Override
+    public String getTileEntityName() {
+        return "gt.multitileentity.multiblock.inventory";
+    }
+
+    @Override
+    public void readMultiTileNBT(NBTTagCompound aNBT) {
+        super.readMultiTileNBT(aNBT);
+        mInventorySize = aNBT.getInteger("mInventorySize");
```
Comment (mitchej123) [inline src/main/java/gregtech/api/multitileentity/multiblock/casing/InventoryUpgrade.java:35]: Don't use strings, put it into a constant in NBT.XXXX

File: src/main/java/gregtech/api/multitileentity/multiblock/casing/InventoryUpgrade.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        mOutputInventory = new ItemStackHandler(mInventorySize);
+
+        loadInventory(aNBT, mInputInventory, NBT.INV_INPUT_LIST);
+        loadInventory(aNBT, mOutputInventory, NBT.INV_OUTPUT_LIST);
+    }
+
+    @Override
+    public void writeMultiTileNBT(NBTTagCompound aNBT) {
+        super.writeMultiTileNBT(aNBT);
+        aNBT.setInteger("mInventorySize", mInventorySize);
```
Comment (mitchej123) [inline src/main/java/gregtech/api/multitileentity/multiblock/casing/InventoryUpgrade.java:47]: shouldn't be needed since it's saved to the NBT from the reference TE

File: src/main/java/gregtech/api/multitileentity/base/BaseMultiTileEntity.java [hunk 8]
```diff
@@ -1009,6 +1010,11 @@ public GT_Packet_New getClientDataPacket() {
                 final ChunkCoordinates aTarget = mtePart.getTargetPos();
                 packet.setTargetPos(aTarget.posX, (short) aTarget.posY, aTarget.posZ);
             }
+            String tName = null;
+            if (this instanceof InventoryUpgrade) {
+                tName = ((InventoryUpgrade) this).getInventoryName();
+            }
+            packet.setInventoryIndex(mtePart.getLockedInventoryIndex(), tName);
```
Comment (mitchej123) [inline src/main/java/gregtech/api/multitileentity/base/BaseMultiTileEntity.java:1017]: should this be set if tName is null or if there is no locked inventory index?

Comment (BlueWeabo) [reply src/main/java/gregtech/api/multitileentity/base/BaseMultiTileEntity.java:1017]: Tho actually. it won't matter if I set it as null or not as i am checking inside the decoding and encoding before hand. and it won't set the any name so its just trying to transfer a null there, which won't be accepted anywhere

Comment (mitchej123) [reply src/main/java/gregtech/api/multitileentity/base/BaseMultiTileEntity.java:1017]: Mainly want to keep packet sizes as small as we can and be sure we only send relevant data when needed.

Comment (BlueWeabo) [reply src/main/java/gregtech/api/multitileentity/base/BaseMultiTileEntity.java:1017]: Ah I see. I can fix that. Will just make a new constant for it for the bit it takes and move over the code.

Comment (BlueWeabo) [reply src/main/java/gregtech/api/multitileentity/base/BaseMultiTileEntity.java]: fixed?

**Review by mitchej123 — APPROVED** [review]
